docs/timeline.md

```diff
@@ -34,9 +31,35 @@ erDiagram
     }
 
     audience 1--0+ audience_member : has
-    actor 1--0+ audience_member : is
-    profile 1--1+ actor : controls
-    share 1--1+ audience : with
-    actor 1--0+ share : initiates
-    share 1+--1 object : refers_to
-```
\ No newline at end of file
+    profile 1--0+ audience_member : is
+    feeds 1--1+ audience : with
+    profile 1--0+ feeds : initiates
+    feeds 1+--1 object : refers_to
+```
+
+## Background
+
+Timelines are the central interaction mechanism for a microblogging service. They need to service extremely high queries rates for relatively large data volumes, as well as constant writes. The returned results need to respect the followings, filters, and lists that are specified for the profile, as well as the creator's and instance's blocks/mutes/etc. As far as I can tell Mastodon (and Calkey/Firefish) accomplishes this by just dumping regular posts into redis and then issuing fairly complex queries to filter down to the correct set of posts to fill out a user's timeline. We're not doing that. Redis is expensive, and scaling it is even more expensive. Also, maybe it's just me, but I think just loading everything into RAM is actually not great system design? And even with the performance of entirely in-memory storage, querying Redis isn't free. Thanks to the hex pattern, changing course and using Redis in the future remains an option, of course.
+
+Instead, we'll be using a timeseries database. Specifically, TimescaleDB, at least to start. Timeseries DBs are analytic—as opposed to transactional—databases. The basic premise is that by giving up on relational integrity, the database can very quickly select, filter, and aggregate very large sets of records. And timeseries DBs in particular, are designed around doing this with data that is ordered and queried chronologically. Exactly like a timeline.
+
+## Details
+
+But, again, there's no relational integrity with any of this. So looking up a post's creator's followers to determine whether it should be excluded by this WHERE is a non-starter. This means we need to do some pre-processing of our data to make it easy to query without those relations. The solution to that is the `Audience` concept. Before they're added to the tsdb, posts get lightly pre-processed to generate audience keys, and these keys essentially constitute feeds, which can be easily merged at query time. Typical audiences would be `public`, the creator's `followers`, and the `mentions` for anyone mentioned in the post. Then to populate a feed, the timeline service will query the feeds table with something like this:
```

Comment:
It seems like this is going to create a really high amount of cardinality in the system (a key per user?), am I understanding that correctly?

Reply:
At least one key per known profile that's posted something since the expiration window, yes. The whole fedi network only has 10m profiles, and only about 2m of them are active. Even if they all were active and followed by members on a single server that's only a cardinality of about 10 million. TBH, I'm not sure how timsecale handles very large cardinality, but I know Influx can handle well into the hundreds of millions, and influx v3 is supposed to be able to handle unlimited cardinality.

Which is to say I'm confident the pattern can scale as far as it will ever need to, even if timescale specifically can't.

Full disclosure, I work at Influx and would have just chosen that, but it fails on the multi-vendor guidance for the project, and doesn't have a great self-hosting story either. I also like that timescale lets small servers operate with just the service and a postgres database.

Reply:
Appreciate that, would love to learn more from you about Time Series in the future!

If I'm understanding things: When someone posts we add a new record keyed with their profile? Are posts going to be keyed by the timeline they are in (ex. a post gets replicated into each users timeline) or by the owner of who posts them?

When I follow a new account or unfollow an account it seems like that could trigger a lot of work if we are creating distinct timelines per user (potentially replicating an entire post history + boosts every follow/unfollow).

Reply:
Keyed based on the creator(s). Specifically the key will be the value for `Profile.FollowersCollection.Id`. And then the followers for that creator will query for that key. So when you follow a new account, that just adds one more item to the WHERE clause to build your home timeline.

And FWIW, I do devops at influx, not like core database stuff. I've picked up some things being in the environment, but I'm not a timeseries expert.

Reply:
Is there a point where adding more WHERE clauses becomes problematic? Will this choice lead to a limitation on follower count for example or degrade performance for users with significant follower counts in a UX impacting way?

Reply:
I don't *think* so. More WHERE clauses could, maybe, depending on the query. But I wouldn't expect more items in the WHERE in collection to cause a problem. Again, I'm running into the limits of my database internals expertise here, but I would expect the query planner to turn that into a hash join, which DBs do all the time.

Reply:
Makes sense, I'm not a DBA either and haven't used that tech. Sorry for all the questions, I do really like this idea though! I can spin this question off and take a look if we want. It seems like this would be a helpful thing to test but I don't think it's productive to block this.

Reply:
No worries at all! Asking questions is exactly what review is for. You're right that some of this is stuff we'll just need to test. I think we're pretty close to being able to do that in a meaningful way.